File: six-degrees-of-kevin-bacon/six_degrees_of_kevin_bacon.py
# Problem:
# https://www.codingame.com/ide/puzzle/six-degrees-of-kevin-bacon

import sys
from collections import deque
from dataclasses import dataclass


@dataclass
class Node:
    value: str
    neighbors: set[str]


def log(*objects):
    # debug-logging in codingame is defined like this.
    print(*objects, file=sys.stderr, flush=True)


def parse_movie_line(line):
    tmp = line.split(":")
    title = tmp[0]
    actors = tmp[1].split(",")
    actors = [actor.strip() for actor in actors]
    return title, set(actors)


def parse_input() -> tuple[str, dict[str, set[str]]]:
    movies: dict[str, set[str]] = dict()

    source_actor = input()
    n = int(input())
    for _ in range(n):
        movie_cast = input()
        title, actors = parse_movie_line(movie_cast)
        movies[title] = actors

    return source_actor, movies


def build_graph(movies) -> dict[str, Node]:
    g: dict[str, Node] = dict()
    for title in movies:
        actors = movies[title]

        for actor in actors:
            neighbors = actors.copy()
            neighbors.remove(actor)

            if actor in g:
                node = g[actor]
                node.neighbors |= neighbors
            else:
                node = Node(value=actor, neighbors=neighbors)
                g[actor] = node
    return g


# dfs alternative
def dfs(g: dict[str, Node], actor, curr_path, shortest_path=sys.maxsize):
    curr_path.append(actor)

    if len(curr_path) < shortest_path:
        if actor == "Kevin Bacon":
            if len(curr_path) < shortest_path:
                log(
                    f"Found new shortest path with length {len(curr_path)}: {curr_path}"
                )

                shortest_path = len(curr_path)
            return shortest_path

        node = g[actor]
        for neighbor in node.neighbors:
            if neighbor not in curr_path:
                tmp = dfs(g, neighbor, [*curr_path], shortest_path)
                shortest_path = min(shortest_path, tmp)

    return shortest_path


# bfs alternative
def bfs(g: dict[str, Node], source_actor: str) -> int:
    bacon_nr = sys.maxsize

    tpl = (source_actor, 1)
    q: deque[tuple[str, int]] = deque()
    q.append(tpl)

    visited = set(source_actor)

    while len(q) > 0:
        tpl = q.popleft()
        curr_actor, dist_from_source = tpl

        # prune long paths
        if dist_from_source > bacon_nr:
            continue

        if curr_actor == "Kevin Bacon":
            bacon_nr = dist_from_source

        node = g[curr_actor]
        for neighbor_name in node.neighbors:
            if neighbor_name not in visited:
                q.append((neighbor_name, dist_from_source + 1))

    return bacon_nr


def get_bacon_nr(source_actor: str, g: dict[str, Node]) -> int:
    # shortest_path = dfs(g, source_actor, [])
    shortest_path = bfs(g, source_actor)

    bacon_nr = shortest_path - 1
    return bacon_nr


if __name__ == "__main__":
    source_actor, movies = parse_input()

    log(source_actor)
    for movie in movies:
        log(f"'{movie}': {movies[movie]},")

    g = build_graph(movies)

    bacon_nr = get_bacon_nr(source_actor, g)

    print(bacon_nr)

File: six-degrees-of-kevin-bacon/six_degrees_of_kevin_bacon.go
package main

import (
	"bufio"
	"container/list"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

type Pair struct {
	first  string
	second int
}

type Node struct {
	value     string
	neighbors map[string]bool
}

func logDebug(format string, args ...interface{}) {
	fmt.Fprintln(os.Stderr, fmt.Sprintf(format, args...))
}

func parseMovieLine(line string) (string, map[string]bool) {
	tmp := strings.Split(line, ":")
	title := tmp[0]
	actors := strings.Split(tmp[1], ",")
	actorSet := make(map[string]bool)
	for _, actor := range actors {
		actorSet[strings.TrimSpace(actor)] = true
	}
	return title, actorSet
}

func parseInput() (string, map[string]map[string]bool) {
	movies := make(map[string]map[string]bool)

	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	sourceActor := scanner.Text()
	scanner.Scan()

	n, _ := strconv.Atoi(scanner.Text())

	for i := 0; i < n; i++ {
		scanner.Scan()
		movieCast := scanner.Text()
		title, actors := parseMovieLine(movieCast)
		movies[title] = actors
	}

	return sourceActor, movies
}

func buildGraph(movies map[string]map[string]bool) map[string]*Node {
	g := make(map[string]*Node)
	for _, actors := range movies {
		for actor := range actors {
			neighbors := make(map[string]bool)
			for neighbor := range actors {
				if neighbor != actor {
					neighbors[neighbor] = true
				}
			}

			if node, ok := g[actor]; ok {
				for neighbor := range neighbors {
					node.neighbors[neighbor] = true
				}
			} else {
				g[actor] = &Node{value: actor, neighbors: neighbors}
			}
		}
	}
	return g
}

func bfs(g map[string]*Node, sourceActor string) int {
	shortestPath := math.MaxInt64

	visited := make(map[string]bool)
	visited[sourceActor] = true

	q := list.New()
	q.PushBack(Pair{sourceActor, 1})

	for q.Len() > 0 {
		e := q.Front()
		q.Remove(e)
		pair := e.Value.(Pair)
		actor := pair.first
		distance := pair.second

		// prune long paths
		if distance > shortestPath {
			continue
		}

		if actor == "Kevin Bacon" {
			shortestPath = distance
		}

		for neighbor := range g[actor].neighbors {
			if !visited[neighbor] {
				visited[neighbor] = true
				q.PushBack(Pair{neighbor, distance + 1})
			}
		}
	}

	return shortestPath
}

func getBaconNr(sourceActor string, g map[string]*Node) int {
	shortestPath := bfs(g, sourceActor)
	baconNr := shortestPath - 1
	return baconNr
}

func main() {
	sourceActor, movies := parseInput()

	//Print parsed input
	logDebug(sourceActor)
	for movie, actors := range movies {
		logDebug("'%s': %v,", movie, actors)
	}

	g := buildGraph(movies)
	baconNr := getBaconNr(sourceActor, g)
	fmt.Println(baconNr)

}
